libqtile/window.py

```diff
@@ -93,6 +93,14 @@
 _NET_WM_STATE_ADD = 1
 _NET_WM_STATE_TOGGLE = 2
 
+# z layers
+NORMAL_LAYER = 0
+ABOVE_LAYER = 1000
+FULLSCREEN_LAYER = ABOVE_LAYER*ABOVE_LAYER
+BELOW_LAYER = -ABOVE_LAYER
+DESKTOP_LAYER = -FULLSCREEN_LAYER
+FLOAT_LAYER = (NORMAL_LAYER + ABOVE_LAYER) / 2
```

Comment:
I would have to check again whether there is a WM_STATE or WM_TYPE hint for that. Currently `FLOAT_LAYER` is just used dynamically when sorting windows, since the fixed value is never assigned. Maybe, at the end of the day, there needs to be a second z-property to differentiate between WM-z and user-defined z.

Reply:
I made the values constant, most importantly because I agree that the ordering looks much nicer.

About changing focus from normal to floating window:
I mostly use the `Bsp` layout, and it doesn't work there, but that might be because that layout implements commands for all the focus changes itself. I did see, however, that the layouts are actually encouraged to specifically allow switching focus into another layout, which would then have to be handled by the group. So there might be something wrong with this either way.
Also, one of the linked issues is about *keeping* floating windows on top, since it is otherwise very easy to "lose" them - which is related to the first issue of not really being able to switch focus between layouts.
There is quite a bit to this topic, but most of it should probably be handled elsewhere.

Reply:
I was thinking if, for example, the user is focussed on a floating window and then clicks (and focusses) a tiled window, presumably the tiled window now appears in front of the floating window?

Reply:
That's explicitly forbidden. In the `Max`-layout (and on occasion every other layout) if a floating window were moved behind a tiling window, there would be no way to get it back to the top. Navigating windows over layout boundaries doesn't work for me (although maybe I have to change my configuration for that?), so I think it is crucial never to hide floating windows.

Reply:
Hmm. I use a keybinding that moves all floating windows to the front so I can continue using them. That seems more flexible than always keeping them in front, which seems opinionated. Should perhaps floating and tiled windows be treated "equal" (and have potentially mixed Z positions), but the behaviour you want be specified using the "always-on-top" functionality we've discussed before? What do you think @tych0 @ramnes ?

Reply:
Even if we wanted to treat them equally, we would have two separate layouts, each suggesting z-levels for its clients but unable to take clients of the other layout into account. So the management done by the layouts would be rendered useless and the group would have to take control again as soon as there is a floating window.

There might be some nuggets in the related issue linked above (#1260 ). A layering mechanism is mentioned in there, and I would favor that.
Since you have a keybinding to work around this: Do you deliberately put your floating windows in the background for later use or is it just something that happened and you needed to come up with a workaround?

Reply:
Looks like I actually got my keybinding from that issue! I deliberately put my floating windows in the background (by focussing another window that happens to be in front of it) and then use the keybinding to get it back.

> we would have two separate layouts, each suggesting z-levels for its clients but unable to take clients of the other layout into account
Right yeah that sounds tricky. I wonder if the behaviour I'm after (and currently use) is simply achieved using the already available `bring_front_click` setting anyway. With that disabled, tiled windows will never move in front of floating windows by following your rules. So maybe this is a non-issue and your way works best :)

Reply:
I wasn't really aware of that setting. Without it, clicked (and/or focused) windows would never change their layer (even within a layout)?

Reply:
Not by clicks, at least, no.

Reply:
That's a bit of a shame. My floating layout raises the focused window at the moment, I'll have to check whether that behaviour is canonical. If not, I'll have to stop it from doing that.

Reply:
I have just been playing around with `bring_front_click` in this branch. With this PR, the setting was useless within a floating layout, since the focused window would always also rise to the top of the float stack. One could, however, click on any tiled window to raise it to the top.

So I tested another approach, where the floating windows are not rearranged along z on focus change (there are two additional commits for this). Now one has to click to raise floating windows again (if `bring_front_click` is true). The only issue is that, as soon as `layout_all` is called, the original z-order is restored (e.g. whenever a new window is added or removed). Fixing that isn't trivial, since the call to raise the window happens in a place where communication with the layout would seem a bit out of place.